erts/emulator/beam/predicates.tab

```diff
@@ -215,27 +215,33 @@ pred.map_key_sort(Size, Rest) {
     return beam_load_map_key_sort(S, Size, Rest);
 }
 
-// Test that the two registers are distinct.
-pred.distinct(Reg1, Reg2) {
-    if (Reg1.type != Reg2.type) {
+// Test that two operands are distinct.
+pred.distinct(Val1, Val2) {
+    if (Val1.type != Val2.type) {
         return 1;
-    } else if (Reg1.type == TAG_x || Reg1.type == TAG_y) {
+    } else if (Val1.type == TAG_x || Val1.type == TAG_y) {
         /* We must not compare the type indices (if any). */
-        return (Reg1.val & REG_MASK) != (Reg2.val & REG_MASK);
+        return (Val1.val & REG_MASK) != (Val2.val & REG_MASK);
+    } else if (Val1.type == TAG_n) {
+        /* NIL has no associated value. */
```

Comment:
Do you think it would be better if we normalized the value of `TAG_n` in `beam_file` instead?

Reply:
Yes, will do.

Reply:
I take it back. I think it's safeer as it is. `beam_file` already sets the value to NIL, but any generator could set the type to `TAG_n` and forget to set the value (or set it to a different value). In fact, [beam_file itself](https://github.com/erlang/otp/blob/07b78d31b1d9edb78aa4a089cc733634ff06e541/erts/emulator/beam/beam_file.c#L1755) sets the type to `TAG_n` without setting the value.